// file-state-monitor/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "FileMonitor", {
  enumerable: true,
  get: function () {
    return _FileMonitor.default;
  }
});
Object.defineProperty(exports, "BaseFileState", {
  enumerable: true,
  get: function () {
    return _states.BaseFileState;
  }
});
Object.defineProperty(exports, "ContentHashState", {
  enumerable: true,
  get: function () {
    return _states.ContentHashState;
  }
});
Object.defineProperty(exports, "LastModifiedState", {
  enumerable: true,
  get: function () {
    return _states.LastModifiedState;
  }
});
Object.defineProperty(exports, "SizeState", {
  enumerable: true,
  get: function () {
    return _states.SizeState;
  }
});
Object.defineProperty(exports, "SmartState", {
  enumerable: true,
  get: function () {
    return _states.SmartState;
  }
});

var _FileMonitor = _interopRequireDefault(require("./FileMonitor"));

var _states = require("./states");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
